webapp/components/notifications/NotificationMenu/NotificationMenu.vue

```diff
@@ -46,24 +48,53 @@ export default {
           variables,
         })
         if (!(markAsRead && markAsRead.read === true)) return
-        this.notifications = this.notifications.map(n => {
-          return n.from.id === markAsRead.from.id ? markAsRead : n
+        this.displayedNotifications = this.displayedNotifications.map(n => {
+          return this.equalNotification(n, markAsRead) ? markAsRead : n
         })
       } catch (err) {
-        throw new Error(err)
+        this.$toast.error(err.message)
       }
     },
+    equalNotification(a, b) {
+      return a.from.id === b.from.id && a.createdAt === b.createdAt && a.reason === b.reason
+    },
   },
   computed: {
-    totalNotifications() {
-      return (this.notifications || []).length
+    notificationsCount() {
+      return (this.displayedNotifications || []).length
+    },
+    unreadNotificationsCount() {
```

Comment:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-26T07:33:23Z" title="Thursday, September 26th 2019, 9:33:23 am +02:00">Sep 26, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -29,13 +41,37 @@ export default {
   },
   data() {
     return {
+      displayedNotifications: [],
       notifications: [],
     }
   },
   props: {
     placement: { type: String },
   },
+  watch: {
+    notifications: {
+      immediate: true,
+      handler(lastQueriedNotifications) {
+        if (lastQueriedNotifications && lastQueriedNotifications.length > 0) {
+          lastQueriedNotifications.forEach(updatedListNotification => {
+            const sameNotification = this.displayedNotifications.find(oldListNotification => {
+              return this.equalNotification(oldListNotification, updatedListNotification)
+            })
+            if (!sameNotification) this.displayedNotifications.unshift(updatedListNotification)
```


Instead of `unshift` I would use `concat` and then `sort` by `createdAt`.

https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/concat
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

Who knows if the notifications arrive in the right order?

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-09-26T07:48:27Z" title="Thursday, September 26th 2019, 9:48:27 am +02:00">Sep 26, 2019</time>_

---


when you say who knows, can you explain that? 
we are currently asking for them back with `orderBy: updatedAt`
doesn't that mean either we do know the order they arrive or something is broken in the functionality of either `neo4j-graphql-js` and/or `neo4j`?

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-09-26T08:35:16Z" title="Thursday, September 26th 2019, 10:35:16 am +02:00">Sep 26, 2019</time>_

---


`concat` in combination with ☝🏼 is a really good idea here @roschaefer . 😘
Is this the preparation for your lecture?

Sorting is not necessary in my eyes.
Should be the right way anyhow and if not in my eyes its better to have the new notifications at the beginning.

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-26T12:48:02Z" title="Thursday, September 26th 2019, 2:48:02 pm +02:00">Sep 26, 2019</time>_

---


I don't know exactly if it could happen in this very example. But since we merge the `read` notifications on the client which are not part of the response that comes from the server (thus not guaranteed to be sorted) I find it good practice to order the list on the client again.

Here's a (very speculative) real world example. Consider we are connected to the fediverse. A notification was created for the user on the other server.

1. Sb. mentions you on a remote server
2. You fetch the notifications
3. The notifications from the remote server arrive at your server
4. You fetch the notifications again. Now the old notification gets put on top even though the `createdAt` is older. This would break the order of the list by `createdAt`.

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-26T13:04:51Z" title="Thursday, September 26th 2019, 3:04:51 pm +02:00">Sep 26, 2019</time>_

---


You don't *have* to implement it. It was an intuition of mine, because we merge something sorted into something not sorted. Feel free to ignore this suggestion

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-09-27T06:48:54Z" title="Friday, September 27th 2019, 8:48:54 am +02:00">Sep 27, 2019</time>_

---


I like your explanation, I hadn't thought of that. I think it doesn't hurt to do it either, I'm sure it's not expensive, especially since the list is more or less sorted anyways

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-09-27T08:48:05Z" title="Friday, September 27th 2019, 10:48:05 am +02:00">Sep 27, 2019</time>_

---


I understand and sort now!